README.md
# Multifunction_Button

"The simplest solutions can solve the most complex questions - that is exactly the purpose of this project!"

Devices are seemingly becoming more and more complex having many instances of needing multiple user input to perform a designated action. This can be seen in the presence of having physical buttons, toggle switches, and even capacitive sensors to detect finger gestures. Generally speaking, a developer tries to make the experience as intuitive and hassle free as possible. Unfortunately this isn't always the case when paired up with an individual who has some kind of physical impairment. Such impairments can completely remove the ability to press a button or flick a switch - Thus rendering the device inoperable by the user. With that being said, this doesn't mean these types of users should be held back from being able to use the device. The barrier lies within the input method to the device and not the users physical limitations.


Two things need to be considered: 
1. What physical abilities does the user have. 
2. What is the best input modality, and how can it be maximized to the fullest potential.

This software assumes that a user has one reliable physical gesture that could be performed multiple times with purposeful intent, as well as the ability to alter the pattern of that gesture to create multiple actions.

Essentially the code monitors a pressed state of an input signal (typically given by a switch of some sort) and calculates the time variance to give two distinct activation gestures - short press, and long press. If the input signal is active within the duration of a short press, then the first action is executed. Vice-versa happens with a long press, executing the second action. 

All the details on how it works are commented within the source files. This code was written for Arduino and is intended to work standalone, as well as be dropped into any project of your choice that needs variance in user control. Download the sketch and all the accompanying source files to go with it. Open inside of the Arduino software or using the specific Visual Studio project files. Compile and flash the code to the microcontroller - or take the foundation of the logic and apply it to another project outside of Arduino.

If you have any questions feel free to contact me!

--ParadoxicalQuad


CHANGES.md
#CHANGES

V2.6:
- Improved Time Calculations.
- Decreased Wait Time For Faster Response.
- Input Pullup (Reversed States).

V2.4:
- Formatted and added comments.
- Used C-style wait function (replaced 'DELAY'), not Arduino reliant.
- Debounce And Edge Detection Are All Included Inside Button Function. 
- Seperated Button function from main code. Implemented in 'button.c' and 'button.h'

V2.3:
- Removed all Debounce for switch filtering.

V2.2:
- Cleaned up code.
- Included third timed gesture.

V2:
- Debounce Switch Filtering.
- Rename variables and improved timed logic.
- Moved Button function implementation back inside the main source.

V1.1:
- Seperated Button function from main code. Implemented in 'button.c' and 'button.h'


button.h
/*
Name:     button.h
Created:  04/15/2017 1:05:00 PM
Updated:  09/26/2017 11:38:00 PM
Author:   Michael Dzura
Version: 2.6

Declaration of variables and functions
*/

#ifndef BUTTON_H
#define BUTTON_H

#include "Arduino.h"

/*********** INPUTS ***********/
#define INPUT_SWITCH	 10	   // Input Pin On Feather 32u4.
/******************************/

/******** TIME VARIANTS *******/
#define CLICK_TIME		 2000   // The Time Of A Single Switch Click. Only Adjust This. Will Multiply By 2.
/******************************/

/****** SWITCH LOGIC **********/
int lastSwitchState;		   // The Previous Reading From The Input Pin.
int pressTrigger;			   // Flag For Switch Press.
int switchState;			   // Value Of The Switch State.
long pressedTime;			   // Timestamp For When The Switch Is Pressed.
long releasedTime;			   // Timestamp For When The Switch Is Released.
long pressDuration;			   // Total Duration The Switch Was Pressed.
/******************************/

/********** DEBOUNCE **********/
#define DEBOUNCEDELAY	 10	   // Debounce Filter Time
long lastDebounceTime;		   // Previous Debounce Timestamp
/******************************/


int button(int input_pin, long short_Click);

#endif


TimeWait.c
/*
Name:     TimeWait.c
Created:  04/15/2017 1:35:00 PM
Author:   Michael Dzura
Version: 1.0

Definition Of TimeWait Function. Creates A Simple Looping 
Pause Based On The Time Parameter Given. This Is To Replace
Arduino Built-in Delay. Use This Library With Any Project 
That Needs Some Kind Of A Delay. This Will Make Your Code  
Less Arduino Dependent.


*/

#include "Arduino.h" // Needs to include Arduino Library. Only Needed When Not Main File.
#include "TimeWait.h" // Includes Variable And Function Declarations.


/************************** Wait Function *******************************
 * Independent Function That Waits In Loop For Defined Interval.		*
 * Similar To Arduino Built-in Delay Function, But Non-Specific			*
 * Change the short_Click variable to adjust the click timing.			*
 ************************************************************************/
void wait(long interval) {
	unsigned long current = 0, previous = 0;
	current = previous = millis();

	// Counts the delay in Milliseconds, and checks for input
	// signal from switch:
	while((current-previous) < interval) {
		current = millis();

		//Serial.println(current-previous);

	}
}


button.c
/*
Name:     button.c
Created:  04/15/2017 1:05:00 PM
Updated:  09/26/2017 11:38:00 PM
Author:   Michael Dzura
Version: 2.6

Definition of button function. Passes input state and press
durations. Activate Short Press (20ms-1000ms), 
Long Press(1000ms-2000ms), Longest Press(2000ms-4000ms).

**TODO
- Double Press functionality.

**CHANGES:
- Debounce And Edge Detection Are All Included Inside Button Function. 
- Improved Time Calculations.
- Decreased Wait Time For Faster Response.
*/

#include "Arduino.h" // Needs to include Arduino Library. Only Needed When Not Main File.
#include "button.h"  // Includes Variable And Function Declarations.


/************************ Button Function *******************************
 * Listens for 3 different gestures from a single switch.				*
 * - Short Click, Long Click, Longest Click.							*
 * Change the short_Click variable to adjust the click timing.			*
 ************************************************************************/
int button(int input_pin, long t_Click) { 

	switchState = digitalRead(input_pin);			// Gets The Realtime Switch State

/*********** Time Calculations *********************
 * Sets Time Calculation Limits					   *
 ***************************************************/
	if (t_Click <= 1000) {
		t_Click = 1000;
	} else if (t_Click >= 4000) {
		t_Click = 4000;
	}


/**************** Edge Detection ********************
 * Check to see if you just pressed the switch		*
 * (i.e. the input went from LOW to HIGH), and		*
 * you've waited long enough since the last press to* 
 * ignore any noise:								*
 *													*
 * For each HIGH/LOW state a timestamp is stored	*
 * representing the rising and falling edge.		*
 ****************************************************/
	if (switchState != lastSwitchState) {

		lastDebounceTime = millis();					// Reset The Debouncing Timer.

		// Check If The Switch Is Being Pressed. If TRUE, Set The Flag.
		if (switchState == LOW) {
			//Serial.print("SwitchState = HIGH, ");
			pressedTime = millis();						// Store The Time Of The Rising Edge.
			//Serial.println(pressedTime);
			
			pressTrigger = 1;							// 1 Indicates That The Switch Has Been Pressed.


		// This is the primer for evaluating every switch press. 
		// It's only true if the button has been previously pressed, 
		// based on the pressTrigger flag, and then released.
		} else if (switchState == HIGH && pressTrigger == 1) {
			//Serial.print("SwitchState = LOW, ");
			releasedTime = millis();					// Store The Time Of The Falling Edge.
			//Serial.println(releasedTime);

			pressDuration = releasedTime - pressedTime;	// Calculate The Total Time The Switch Was Pressed

			//Serial.println("");
			//Serial.print("Switch Press For:  ");
			//Serial.println(pressDuration);

			pressTrigger = 0;							// Reset the press trigger back to zero, not pressed.

			/**************************** Short Click Perameters****************************
			* Greater Than Debounce Filter, And Less Than Short Click Time.			   *
			*******************************************************************************/
			if ((pressDuration > DEBOUNCEDELAY) && (pressDuration <= t_Click) && pressTrigger == 0) {
				lastDebounceTime = millis();			// Reset Debounce Timer.
				//Serial.println("Short Press!");
				pressTrigger = 0;						// Reset the press trigger back to zero.
														// Prevents Release From Activating.
				return 1;								// Return If True.
			}

			/**************************** Long Click Perameters ****************************
			* Greater Than Short Click Time, And Less Than Long Click Time.			   *
			*******************************************************************************/
			if ((pressDuration > t_Click) && (pressDuration <= (t_Click * 2)) && pressTrigger == 0) {
				lastDebounceTime = millis();			// Reset Debounce Timer.
				//Serial.println("Long Press!");
				pressTrigger = 0;						// Reset the press trigger back to zero.
														// Prevents Release From Activating.
				return 2;								// Return If True.
			} 

			/*************************** Longest Click Perameters **************************
			* Greater Than Long Click Time, And Less Than Longest Click Time.			   *
			* Timeout Is Included For Anything Else Greater.							   *
			*******************************************************************************/
			if ((pressDuration > (t_Click * 2)) && (pressDuration < (t_Click * 3)) && pressTrigger == 0) {
				lastDebounceTime = millis();			// Reset Debounce Timer.
				//Serial.println("Longest Press!");
				pressTrigger = 0;						// Reset the press trigger back to zero.
														// Prevents Release From Activating.
				return 3;								// Return If True.
			}
		}
	}

	// Save the switch value. Next time through the loop,
	// it'll be the lastButtonState:
	lastSwitchState = switchState;

	return 0; // No Action Returns Nothing.
}
